src/batch/src/executor2/sort_agg.rs

```diff
@@ -0,0 +1,635 @@
+
+use crate::executor::ExecutorBuilder;
+use crate::executor2::{BoxedDataChunkStream, BoxedExecutor2, BoxedExecutor2Builder, Executor2};
+
+/// `SortAggExecutor` implements the sort aggregate algorithm, which assumes
+/// that the input chunks has already been sorted by group columns.
+/// The aggregation will be applied to tuples within the same group.
+/// And the output schema is `[group columns, agg result]`.
+///
+/// As a special case, simple aggregate without groups satisfies the requirement
+/// automatically because all tuples should be aggregated together.
+pub struct SortAggExecutor2 {
+    agg_states: Vec<BoxedAggState>,
+    group_keys: Vec<BoxedExpression>,
+    sorted_groupers: Vec<BoxedSortedGrouper>,
+    child: BoxedExecutor2,
+    schema: Schema,
+    identity: String,
+    output_size_limit: usize,
+}
+
+impl BoxedExecutor2Builder for SortAggExecutor2 {
+    fn new_boxed_executor2(source: &ExecutorBuilder) -> Result<BoxedExecutor2> {
+        ensure!(source.plan_node().get_children().len() == 1);
+        let proto_child =
+            source.plan_node().get_children().get(0).ok_or_else(|| {
+                ErrorCode::InternalError("SortAgg must have child node".to_string())
+            })?;
+        let child = source.clone_for_plan(proto_child).build2()?;
+
+        let sort_agg_node = try_match_expand!(
+            source.plan_node().get_node_body().unwrap(),
+            NodeBody::SortAgg
+        )?;
+
+        let agg_states = sort_agg_node
+            .get_agg_calls()
+            .iter()
+            .map(|x| AggStateFactory::new(x)?.create_agg_state())
+            .collect::<Result<Vec<BoxedAggState>>>()?;
+
+        let group_keys = sort_agg_node
+            .get_group_keys()
+            .iter()
+            .map(build_from_prost)
+            .collect::<Result<Vec<BoxedExpression>>>()?;
+
+        let sorted_groupers = group_keys
+            .iter()
+            .map(|e| create_sorted_grouper(e.return_type()))
+            .collect::<Result<Vec<BoxedSortedGrouper>>>()?;
+
+        let fields = group_keys
+            .iter()
+            .map(|e| e.return_type())
+            .chain(agg_states.iter().map(|e| e.return_type()))
+            .map(Field::unnamed)
+            .collect::<Vec<Field>>();
+
+        Ok(Box::new(Self {
+            agg_states,
+            group_keys,
+            sorted_groupers,
+            child,
+            schema: Schema { fields },
+            identity: source.plan_node().get_identity().clone(),
+            output_size_limit: DEFAULT_CHUNK_BUFFER_SIZE,
+        }))
+    }
+}
+
+impl Executor2 for SortAggExecutor2 {
+    fn schema(&self) -> &Schema {
+        &self.schema
+    }
+
+    fn identity(&self) -> &str {
+        &self.identity
+    }
+
+    fn execute(self: Box<Self>) -> BoxedDataChunkStream {
+        self.do_execute()
+    }
+}
+
+impl SortAggExecutor2 {
+    /// We assume output_size_limit() will return the size limit of child chunks,
```

Comment:
In current implementation this may be false, some executors in doesn't follow this.

Reply:
Here I think after we finish refactoring all executors, the assumption can be hold, am I right？

Reply:
Yes, but it maybe a long process and may causes problems during it.